### PR TITLE
HPCC-14802 Add additional edit distance metrics

### PR DESCRIPTION
**Pull request by jamienoss**
Add both the restricted (optimal string alignment - OSA) and unrestricted (full)
Damerau-Levenshtein metrics.
Add a weighted version of the full Damerau-Levenshtein.

NOTE: this doesn't currently contain the service entry for passing user
defined weight-matrices nor wrappers for passing dataset/tables etc.

Signed-off-by: James Noss james.noss@lexisnexis.com

### CODE REVIEW

**Comment by jamienoss** [comment]
@mckellyln could you take a look please? Perhaps we can add this and then have it replaced with your optimised implementation when it's ready?

**Comment by mckellyln** [comment]
ok, sure.
My routine checks if 2 strings are OSA within a provided radius and only if 1 <= radius <= 2.
So I think we need a routine that is something like WithinOSARadius(string s1, string s2, int radius)
and in it check if radius == 0 then just string compare, if radius > 2 then run pure OSA and return if distance found is <= radius, otherwise call my routine.

File: ecllibrary/teststd/str/TestDamerauLevenshteinDistance.ecl [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  STRING manyAlpha := alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha;
+  STRING manyDigits := digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+
+                       digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits;
+
+  EXPORT TestConst := MODULE
+    EXPORT Test01 := ASSERT(Str.DamerauLevenshteinDistance('','') = 0, CONST);
+    EXPORT Test02 := ASSERT(Str.DamerauLevenshteinDistance('','                ') = 0, CONST);
+    EXPORT Test03 := ASSERT(Str.DamerauLevenshteinDistance('                ','') = 0, CONST);
+    EXPORT Test04 := ASSERT(Str.DamerauLevenshteinDistance('a ','                ') = 1, CONST);
+    //EXPORT Test05 := ASSERT(Str.DamerauLevenshteinDistance(' a ','   ') = 1, CONST);
```
Comment (ghalliday) [inline ecllibrary/teststd/str/TestDamerauLevenshteinDistance.ecl]: Should this be removed or uncommented?

File: ecllibrary/teststd/str/TestWeightedDamerauLevenshteinDistance.ecl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    EXPORT Test06 := ASSERT(Str.WeightedDamerauLevenshteinDistance('Aprs  ','APp') = 2, CONST);//default confusion matrix is case insensitive.
+    EXPORT Test07 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','acbd') = 1, CONST);
+    EXPORT Test08 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','abd') = 1, CONST);
+    EXPORT Test09 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','abc') = 1, CONST);
+    EXPORT Test10 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','bcd') = 1, CONST);
+    EXPORT Test11 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','abcde') = 1, CONST);
+    EXPORT Test12 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','aabcd') = 1, CONST);
+    EXPORT Test13 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd',' abcd') = 1, CONST);
+    EXPORT Test14 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','a bcd') = 1, CONST);
+    EXPORT Test15 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','adcd') = 0, CONST);
```
Comment (ghalliday) [inline ecllibrary/teststd/str/TestWeightedDamerauLevenshteinDistance.ecl]: Curious.  Is this useful if the function returns 0 for non-identical strings?

File: plugins/stringlib/stringlib.cpp [hunk 3]
```diff
@@ -223,6 +229,269 @@ unsigned editDistance(unsigned leftLen, const char * left, unsigned rightLen, co
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+unsigned optimalStringAlignmentDistance(unsigned leftLen, const char * left, unsigned rightLen, const char * right)
```
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: trivial.  This can be a static function.

File: plugins/stringlib/stringlib.cpp [hunk 4]
```diff
@@ -190,7 +196,7 @@ unsigned editDistance(unsigned leftLen, const char * left, unsigned rightLen, co
     //Optimize the storage requirements by
     //i) Only storing two stripes
     //ii) Calculate, but don't store the row comparing against the null string
-    unsigned char da[2][256];
+    unsigned char da[3][256];
```
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: Was this change intentional?

File: plugins/stringlib/stringlib.cpp [hunk 5]
```diff
@@ -223,6 +229,269 @@ unsigned editDistance(unsigned leftLen, const char * left, unsigned rightLen, co
...
 }
 
+unsigned optimalStringAlignmentDistance(unsigned leftLen, const char * left, unsigned rightLen, const char * right)
+{
+    /* This is the restricted form of the Damerau-Levenshtein distance AKA optimal string alignment (OSA).
+     * It is restricted in that a substring can only be altered once.
+     * e.g. OSA(CA, ABC) = 3, CA -> A -> AB -> ABC : fullDL(CA, ABC) = 2, CA -> AC -> ABC (i.e. a transposition then an insertion upon that transposition).
+     * e.g. OSA(abcde, acbed) = 3 & fullDL(abcde, acbed) = 2
+     */
+    if (rightLen < 2 || leftLen < 2)
```
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: minor: Should this be done after trimming?

File: plugins/stringlib/stringlib.cpp [hunk 6]
```diff
@@ -223,6 +229,269 @@ unsigned editDistance(unsigned leftLen, const char * left, unsigned rightLen, co
...
+    unsigned char da[3][256];
+    char r_0 = right[0];
+    char l_0 = left[0];
+    bool matched_l0 = false;
+    //Transpositions aren't included until i,j > 1 to allow for -2 offset - therefore this first row is the same as for Levenshtein.
+    for (j = 0; j < rightLen; j++)
+    {
+        if (right[j] == l_0) matched_l0 = true;
+        da[0][j] = (matched_l0) ? j : j+1;
+        da[2][j] = j; //This is the NULL row. NOTE: da[2][reightLen] is never actually referenced.
```
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: trivial: rightLen in comment

File: ecllibrary/teststd/str/TestWeightedDamerauLevenshteinDistance.ecl (hunk 2 above)
Comment (jamienoss) [inline ecllibrary/teststd/str/TestWeightedDamerauLevenshteinDistance.ecl]: It is due to the weight matrix in that 'b' & 'd' can be optically confused and thus have a zero weight.

So the other three distance metrics give a distance of 1 for these two strings and thus to differentiate with the weighted version the only option is for the operation to be ignored and thus yield 0.

It is a good reason to expand to allow user defined weight matrices.

More importantly thank you for noting this as I had forgotten about it and assumed, as you did, for the auto-id-suggestion code, that d = 0 => identical-identical rather than weighted-identical.

File: ecllibrary/teststd/str/TestDamerauLevenshteinDistance.ecl (hunk 1 above)
Comment (jamienoss) [inline ecllibrary/teststd/str/TestDamerauLevenshteinDistance.ecl]: Uncommented and changed to = 2.

File: plugins/stringlib/stringlib.cpp (hunk 4 above)
Comment (jamienoss) [inline plugins/stringlib/stringlib.cpp]: No, good spot.

File: plugins/stringlib/stringlib.cpp (hunk 5 above)
Comment (jamienoss) [inline plugins/stringlib/stringlib.cpp]: yep, thanks.

File: plugins/stringlib/stringlib.cpp (hunk 3 above)
Comment (jamienoss) [inline plugins/stringlib/stringlib.cpp]: I'll leave as is, I haven't decided if using the weighted version is the best choice for the auto id suggestions, I running bulk tests to compare them all but needed to stop when you reminded me that weightedDistance(str1, str2) == 0 !=> strcmp(str1, str2) == 0.

**Comment by jamienoss** [comment]
@ghalliday Thanks, have made changes.

File: plugins/stringlib/stringlib.cpp [hunk 7]
```diff
@@ -223,6 +229,270 @@ unsigned editDistance(unsigned leftLen, const char * left, unsigned rightLen, co
...
+    if (rightLen > 255)
+        rightLen = 255;
+
+    if (leftLen == 0)
+        return rightLen;
+
+    if (rightLen == 0)
+        return leftLen;
+
+#if 0
```
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: This code should be deleted, or moved/simplified in a comment if it is being retained for reference.

File: plugins/stringlib/stringlib.cpp (hunk 3 above)
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: I meant prefix with the keyword "static" since it isn't exported out side of this module.

**Comment by ghalliday** [comment]
@jamienoss a couple of minor issues.
@mckellyln can you (relatively briefly) sanity check the code.

**Comment by jamienoss** [comment]
@ghalliday changes made.
@mckellyln - mentioning you so PR summary email keeps you as the owner.

**Comment by mckellyln** [comment]
I compared this new damerauLevenshteinDistance to my reference OSA and get expected results, with damerauLevenshteinDistance finding only more matches than my OSA, such as:
SIMLEY             SIMERLY

But comparing this new optimalStringAlignmentDistance to my reference OSA I see some differences I do not agree with, such as:
SICKLES            STICKELS - my OSA finds this and optimalStringAlignmentDistance does not
SIDDIQ               UDDIN - optimalStringAlignmentDistance finds this but my OSA does not

**Comment by mckellyln** [comment]
This is my reference OSA:
```
    for (i=0;i<=lenL;i++)
        mat[i][0] = i;

    for (j=0;j<=lenR;j++)
        mat[0][j] = j;

    cost = 0;
    for (j = 1; j <= lenR; j++) {
        mind = edis + 2;
        for (i = 1; i <= lenL; i++) {
            cost = (l[i-1] == r[j-1]) ? 0 : 1;
            mat[i][j] = MXMIN(mat[i][j-1] + 1, MXMIN(mat[i-1][j] + 1, mat[i-1][j-1] + cost));
            if(i > 1 && j > 1 && l[i-1] == r[j-2] && l[i-2] == r[j-1]) {
                mat[i][j] = MXMIN(mat[i][j], mat[i-2][j-2] + cost);
            }
            mind = MXMIN(mind, mat[i][j]);
        }
        if(mind > edis)
            return false;
    }

    if (mat[lenL][lenR] <= edis)
        return true;
```

**Comment by mckellyln** [comment]
An an example, here is a complete version of code for:
OSA
OSAWithinRadius
which also uses optimal performing version for when radius <= 2

`static unsigned optimalStringAlignmentDistanceWithinRadiusAny(unsigned leftLen, const char * left, unsigned rightLen, const char * right, unsigned radius, bool useWithin=true)
{
    int w = 1;
    int s = 1;
    int a = 1;
    int d = 1;
    unsigned char lrow0[256];
    unsigned char lrow1[256];
    unsigned char lrow2[256];
    unsigned char *row0;
    unsigned char *row1;
    unsigned char *row2;
    unsigned int i, j;
    unsigned mind;

    clip(leftLen, left);
    clip(rightLen, right);

    if (leftLen > 255)
        leftLen = 255;

    if (rightLen > 255)
        rightLen = 255;

    if (leftLen == 0)
        return rightLen;

    if (rightLen == 0)
        return leftLen;

    row0 = &lrow0[0];
    row1 = &lrow1[0];
    row2 = &lrow2[0];

    for (j = 0; j <= rightLen; j++)
        row1[j] = j * a;

    for (i = 0; i < leftLen; i++)
    {
        unsigned char *dummy;
        row2[0] = (i + 1) * d;

        mind = 256;

        for (j = 0; j < rightLen; j++)
        {
            /* substitution */
            row2[j + 1] = row1[j] + s * (left[i] != right[j]);

            /* swap */
            if (i > 0 && j > 0 && left[i - 1] == right[j] && left[i] == right[j - 1] && row2[j + 1] > row0[j - 1] + w)
                row2[j + 1] = row0[j - 1] + w;

            /* deletion */
            if (row2[j + 1] > row1[j + 1] + d)
                row2[j + 1] = row1[j + 1] + d;

            /* insertion */
            if (row2[j + 1] > row2[j] + a)
                row2[j + 1] = row2[j] + a;

            if (useWithin)
                mind = MXMIN(mind, row2[j+1]);
        }

        if (useWithin && mind > radius)
            return 0;

        dummy = row0;
        row0 = row1;
        row1 = row2;
        row2 = dummy;
    }

    if (!useWithin)
        return row1[rightLen];

    if(row1[rightLen] <= radius)
        return row1[rightLen];

    return 0;
}

unsigned optimalStringAlignmentDistance(unsigned leftLen, const char * left, unsigned rightLen, const char * right)
{
    return optimalStringAlignmentDistanceWithinRadiusAny(leftLen, left, rightLen, right, 256, false);
}

typedef struct {
    char method[2];
} EDIT_LIST;

static bool optimalStringAlignmentDistanceWithinRadius2(unsigned leftLen, const char * left, unsigned rightLen, const char * right, int radius)
{
    /*
     * Optimal String Alignment (OSA) string compare function
     * (also referred to as 'restricted' Damerau-Levenshtein)
     *
     * Adaption to C with optimizations and additional code for transpositions and 
     * trailing half-edits by Mark Kelly @ LexisNexis, June 2015
     *
     * Original python work from Fujimoto Seiji:
     * Copyright (c) 2012 Fujimoto<fujimoto@writingarchives.sakura.ne.jp>
     * 
     * Permission is hereby granted, free of charge, to any person obtaining a copy
     * of this software and associated documentation files (the "Software"), to deal
     * in the Software without restriction, including without limitation the rights
     * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     * copies of the Software, and to permit persons to whom the Software is furnished
     * to do so, subject to the following conditions:
     * 
     * The above copyright notice and this permission notice shall be included in all
     * copies or substantial portions of the Software.
     */

    bool incl_half_edits = false;

    clip(leftLen, left);
    clip(rightLen, right);

    if (leftLen > 255)
        leftLen = 255;

    if (rightLen > 255)
        rightLen = 255;

    if (leftLen == 0)
        return rightLen;

    if (rightLen == 0)
        return leftLen;

    int l1 = leftLen;
    int l2 = rightLen;

    unsigned char *str1 = (unsigned char *)left;
    unsigned char *str2 = (unsigned char *)right;

    if (l1 < l2) {
        str1 = (unsigned char *)right;
        str2 = (unsigned char *)left;
        int Ltmp = l1;
        l1 = l2;
        l2 = Ltmp;
    }

    char SUB = 's';
    char INS = 'i';
    char DEL = 'd';
    char TRN = 't';
    char NUL = 'x';

    EDIT_LIST edit_list[12];
    int num_lists = 0;

    int ldif = l1 - l2;

    if (radius == 0) {
        if (ldif == 0) {
            edit_list[num_lists].method[0] = NUL;
            edit_list[num_lists].method[1] = NUL;
            num_lists += 1;
        } else
            return false;
    } else if (radius == 1) {
        if (ldif == 0) {
            edit_list[num_lists].method[0] = SUB;
            edit_list[num_lists].method[1] = NUL;
            num_lists += 1;
            edit_list[num_lists].method[0] = TRN;
            edit_list[num_lists].method[1] = NUL;
            num_lists += 1;
        } else if (ldif == 1) {
            edit_list[num_lists].method[0] = DEL;
            edit_list[num_lists].method[1] = NUL;
            num_lists += 1;
        } else if (!incl_half_edits)
            return false;
        else {
            if (ldif == 2) {
                edit_list[num_lists].method[0] = DEL;
                edit_list[num_lists].method[1] = NUL; // 1 remain with 1 edit
                num_lists += 1;
                edit_list[num_lists].method[0] = SUB;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 1 edit
                num_lists += 1;
                edit_list[num_lists].method[0] = TRN;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 1 edit
                num_lists += 1;
                edit_list[num_lists].method[0] = NUL;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 0 edits
                num_lists += 1;
            } else if (ldif == 3) {
                edit_list[num_lists].method[0] = DEL;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 1 edit
                num_lists += 1;
            } else if (ldif == 4) {
                edit_list[num_lists].method[0] = NUL;
                edit_list[num_lists].method[1] = NUL; // 4 remain with 0 edits
                num_lists += 1;
            } else
                return false;
        }
    } else if (radius == 2) {
        if (ldif == 0) {
            edit_list[num_lists].method[0] = INS;
            edit_list[num_lists].method[1] = DEL;
            num_lists += 1;
            edit_list[num_lists].method[0] = DEL;
            edit_list[num_lists].method[1] = INS;
            num_lists += 1;
            edit_list[num_lists].method[0] = SUB;
            edit_list[num_lists].method[1] = SUB;
            num_lists += 1;
            edit_list[num_lists].method[0] = TRN;
            edit_list[num_lists].method[1] = TRN;
            num_lists += 1;
            edit_list[num_lists].method[0] = TRN;
            edit_list[num_lists].method[1] = SUB;
            num_lists += 1;
            edit_list[num_lists].method[0] = SUB;
            edit_list[num_lists].method[1] = TRN;
            num_lists += 1;
        } else if (ldif == 1) {
            edit_list[num_lists].method[0] = DEL;
            edit_list[num_lists].method[1] = SUB;
            num_lists += 1;
            edit_list[num_lists].method[0] = SUB;
            edit_list[num_lists].method[1] = DEL;
            num_lists += 1;
            edit_list[num_lists].method[0] = DEL;
            edit_list[num_lists].method[1] = TRN;
            num_lists += 1;
            edit_list[num_lists].method[0] = TRN;
            edit_list[num_lists].method[1] = DEL;
            num_lists += 1;
        } else if (ldif == 2) {
            edit_list[num_lists].method[0] = DEL;
            edit_list[num_lists].method[1] = DEL;
            num_lists += 1;
            if (incl_half_edits) {
                edit_list[num_lists].method[0] = SUB;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 1 edit
                num_lists += 1;
                edit_list[num_lists].method[0] = TRN;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 1 edit
                num_lists += 1;
            }
        } else if (!incl_half_edits)
            return false;
        else {
            if (ldif == 3) {
                edit_list[num_lists].method[0] = DEL;
                edit_list[num_lists].method[1] = NUL; // 2 remain with 1 edit
                num_lists += 1;
                edit_list[num_lists].method[0] = NUL;
                edit_list[num_lists].method[1] = NUL; // 3 remain with 0 edits
                num_lists += 1;
            } else if (ldif == 4) {
                edit_list[num_lists].method[0] = NUL;
                edit_list[num_lists].method[1] = NUL; // 4 remain with 0 edits
                num_lists += 1;
            } else
                return false;
        }
    }

    int emin = 3;
    for (int k=0; k<num_lists; k++) {
        char *method = edit_list[k].method;
        int i = 0, j = 0;
        int d = 0;
        while (i < l1 && j < l2) {
            if (__builtin_expect((str1[i] != str2[j]), 1)) {
                d += 1;
                if (d > radius)
                    break;
                char emeth = method[d-1];
                if (emeth == DEL) 
                    i += 1;
                else if (emeth == INS)
                    j += 1;
                else if (emeth == SUB) {
                    i += 1;
                    j += 1;
                } else if (emeth == TRN) {
                    if (__builtin_expect(( (i < l1-1) && (j < l2-1) &&
                            (str1[i+1] == str2[j]) && (str1[i] == str2[j+1]) ),0)) {
                        i += 2;
                        j += 2;
                    } else {
                        if ( (incl_half_edits) && (j == l2-1) && (l1-i-1 <= 2*(radius-d)) )
                            return true;
                        else {
                            d = 3;
                            break;
                        }
                    }
                }
            } else {
                i += 1;
                j += 1;
            }
        }

        if (__builtin_expect((d > radius), 1)) {
            continue;
        } else if (i < l1) {
            if (incl_half_edits) {
                if (l1-i <= 2*(radius-d))
                    return true;
                else
                    continue;
            } else {
                int eleft = 0;
                if (d == 0)
                    eleft = (method[d] == DEL) + (method[d+1] == DEL);
                else if (d == 1)
                    eleft = (method[d] == DEL);
                if (l1-i <= eleft)
                    d += (l1-i);
                else
                    continue;
            }
        } else if (j < l2) {
            if (incl_half_edits) {
                if (l2-j <= 2*(radius-d))
                    return true;
                else
                    continue;
            } else {
                int eleft = 0;
                if (d == 0)
                    eleft = (method[d] == INS) + (method[d+1] == INS);
                else if (d == 1)
                    eleft = (method[d] == INS);
                if (l2-j <= eleft)
                    d += (l2-j);
                else
                    continue;
            }
        }
        if (d < emin)
            emin = d;
        if (emin <= radius)
            return true;
    }
    if (emin <= radius)
        return true;

    return false;
}

bool optimalStringAlignmentDistanceWithinRadius(unsigned leftLen, const char * left, unsigned rightLen, const char * right, unsigned radius)
{
    if (radius <= 2)
        return optimalStringAlignmentDistanceWithinRadius2(leftLen, left, rightLen, right, radius);
    else
        return (bool)optimalStringAlignmentDistanceWithinRadiusAny(leftLen, left, rightLen, right, radius, true);
}
`
